python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+
+```python
+class Client:
+    def __init__(
+        self,
+        addr: str,
+        *,
+        namespace: str = "default",
+        default_task_queue: str = None,
+        identity: str = None
+        # ... additional options omitted for brevity
+    ) -> None:
+        # Actually dials gRPC here
+        pass
+
+    # Raw gRPC service
+    @property
+    def service(self) -> workflowservice.WorkflowService:
+        pass
+
+    def start_workflow(
+        workflow: str,
+        *args,
+        id: str = None,
+        task_queue: str = None,
+        retry_policy: RetryPolicy = None,
+        # Note, we are embedding signal-with-start in here
+        start_signal: str = None,
+        start_signal_args=[],
+        # ... additional options omitted for brevity
+    ) -> WorkflowHandle[Any]:
+        pass
+
+    # In a future phase:
+    # @overload
+    # def start_workflow(workflow: Callable[P, T], *args: P.args, ...) -> WorkflowExecution[T]
+
+    # Does not validate whether a workflow exists
+    def workflow(
+        workflow_id: str,
+        run_id: str = None
+        # ...
+    ) -> WorkflowHandle[Any]:
+        pass
+
+
+class WorkflowHandle(Generic[T]):
+    def __init__(
+        self,
+        client: Client,
+        workflow_id: str,
+        *,
+        run_id: str = None
+        # ...
+    ) -> None:
+        pass
+
+    def __await__(self) -> T:
+        pass
+
+    async def cancel(self):
+        pass
+
+    async def describe(self):
+        pass
+
+    async def query(self, name: str, *args):
+        pass
+
+    async def signal(self, name: str, *args):
+        pass
+
+    async def terminate(self, *, reason: str = None):
+        pass
+```
+
+Notes:
+
+* We are intentionally defining a `Client` instead of a `WorkflowClient` since this will be general purpose
+  * Having a general purpose client keeps things simple, but at least for now I can't see defining any more than the two
+    methods and the one property
+  * We could make an abstract base class if we want for typing/purposes that just has the `service` property
+    unimplemented
+* We call it `start_workflow` instead of just `start` to be clearer
+* We have embedded `start_signal` and `start_signal_args` as parameters, so if the former is present we will perform a
+  signal with start. Is this ok?
+  * This is an intentional deviation from other SDKs, but makes logical sense
+* The client supports a default task queue, do we want this?
+
+### Worker
+
+Exported from `temporalio` (but probably in `temporalio/worker/worker.py`, exported from `temporalio.worker` then
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        client: temporalio.Client,
+        *,
+        # Uses default in client if not present (or errors if no default)
+        task_queue: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
```

Comment:
Maybe instead of all of these `register_activity` calls we can just give a mapping of name to implementation to the worker at construction time, then you don't have to worry about when this method can be called.

That way you can just do something like:
```python
import activities

Worker(task_queue, activities=vars(activities))
```

Reply:
I brought up this point above with:

> Should we accept a set of activities to register in `__init__`?

Somewhat discussed at https://github.com/temporalio/proposals/pull/49/files/40c1178e1937b223c93334a7cefc75e31afc6bb9#r792182996. I'm now thinking both. I don't want the worker to start on construction anyways (I need it awaitable), so `run()` will always exist. Also, there benefit of the separate `register_` functions with regards to overload typing, documentation, and potential future per-activity options.

Reply:
I like that we have `Worker.create()` and `Worker.run()` in TS.
Note that `create()` will try to connect to the server.

Reply:
For the `register_*` functions, there are other ways to merge multiple activity sources without them.

Reply:
I added the ability to provide them at creation. Since we are creating core separately from worker, we don't need a static async `create`, class init is fine. The `run` here is like the one in TS.

Reply:
If Core is a parameter we don't need an async create method. We will need one to support a singleton Core.